mlir/lib/Dialect/Affine/Transforms/doc/AffineLoopInterchange.md

```diff
@@ -0,0 +1,136 @@
+synchronization operations are more costly on multicores due to cache coherence 
+policies.
+
+The costs are calculated as follows:
+
+#### Spatial Locality Cost:
+
+The Spatial locality cost is based on size of the locality of each 
+forLoop. Some define it as number of cahce lines accessed in this for loop were 
+the innermost loop. A method for calculating these values are given in a paper
+by Steve Carr et al. [here](https://dl.acm.org/doi/abs/10.1145/195470.195557).
+
+Once the locality information is available, we define the cost of a given 
+permutation as the sum of product of locality of each loop multiplied with the 
+size of iteration sub-space defined by the loops inside that loop in this
+permutation. That is, `Cost_Permutation = Sum_over_all_loop(locality of loop i * 
+Size of iteration sub-space of all loops which are defined inside loop i`
+
+For a loopnest l1, l2, l3 this will be equal to 
+Cost_l1_l2_l3 = locality_l1* number of iterations in l2* number of iterations in 
+l3 + locality_l2 * number of iterations in l3 + locality_l3.
+
+###### Please note that this is just a representative cost and not some 
+absolute quantity.
+
+#### Temporal Reuse Cost:
+
+The Temporal Reuse cost is based on the sum total number of reuses experienced
+by each memory access for a given permutation. For each permutation, we check 
+for each loop (starting from the innermost loop) if the column corresponding to
+that loop's IV is all zeros in the access matrix of any given load/store. If
+the column is all zeros, it means the same value is accessed over the entire 
+loop iteration - hence the reuse count increases by a factor of O(n). The same 
+procedure is repeated till we encounter a loop for which the column in access 
+matrix is non-zero. The reuse count stops there. No matter whether further
+loops have a all-zero column or not, the reuse will not take place once a non- 
+zero column has been encountered.
+
+Based on the total reuse count, we assign a cost to the permutation such that 
+maximal temporal reuse corresponds to a minimum cost.
+
+This approach works since over all the permutation, the one which allows for 
+maximum temporal reuse will receive minimum cost - even if spatial costs are 
+same, the permutation having high temporal reuse will be selected over the one 
+having low temporal reuse.
+
+#### Parallelism Cost:
+
+For parallelism, we make use of loop-carried-dependence vector. We define 
+individual loopcost as the size of iteration sub-space (product of iteration 
+count of underlying loops) if the loop carries a dependency. Otherwise, the 
+individual loopcost for this loop is zero.
+
+The final cost of a permutation is defined as the sum over all the individual 
+loopcosts.This approach works because it favours the permutations which have 
+more number of free outer loops.
+
+
+#### Conclusion:
+
+Thus, the permutation to be selected as the best permutation is depends on all - 
```

Comment:
Grammatically broken -> ".. is depends on all"

Reply:
Fixed. Thanks.